icu4c/source/i18n/smpdtfmt.cpp

```diff
@@ -2575,51 +2582,47 @@ SimpleDateFormat::parse(const UnicodeString& text, Calendar& cal, ParsePosition&
             } else { // tztype == TZTYPE_DST
                 if (dst == 0) {
                     if (btz != NULL) {
-                        UDate time = localMillis + raw;
-                        // We use the nearest daylight saving time rule.
-                        TimeZoneTransition beforeTrs, afterTrs;
-                        UDate beforeT = time, afterT = time;
-                        int32_t beforeSav = 0, afterSav = 0;
-                        UBool beforeTrsAvail, afterTrsAvail;
-
-                        // Search for DST rule before or on the time
-                        while (TRUE) {
-                            beforeTrsAvail = btz->getPreviousTransition(beforeT, TRUE, beforeTrs);
-                            if (!beforeTrsAvail) {
+                        // This implementation resolves daylight saving time offset
+                        // closest rule after the given time.
+                        UDate baseTime = localMillis + raw;
+                        UDate time = baseTime;
+                        UDate limit = baseTime + MAX_DAYLIGHT_DETECTION_RANGE;
+                        TimeZoneTransition trs;
+                        UBool trsAvail;
+
+                        // Search for DST rule after the given time
+                        while (time < limit) {
```

Comment:
Small improvement. If there is a zone alter multiple offsets infinitely in future, and neither of them are not daylight saving time, while(TRUE) loop won't terminate. Updated implementation uses ~30 year window, and does not look for further. I'm sure there no such rules available at this moment although.